README.md
# canary.sh

Pure Bash vanilla Kubernetes Canary rollouts

--------

## Installation

Get the script, put it somewhere in PATH, and make it executable. Example:

```bash
curl -sSL https://git.io/canary.sh -o /usr/local/bin/canary.sh
# Always verify the contents of anything you curl down before running!
less /usr/local/bin/canary.sh
chmod +x /usr/local/bin/canary.sh
canary.sh -h
```

You can also get specific tags or commits. Example:

```bash
curl -sSL https://raw.githubusercontent.com/zacanger/canary.sh/v0.4.0/canary.sh -o canary.sh
```

See the [changelog](./CHANGELOG.md) for history.

## Prerequisites

* Bash 4+
* kubectl
* GNU sed (if you have both `sed` and `gsed`, the script will use `gsed`)
* An existing deployment and service. These will need to be modified slightly
  to work with this script. See the example below for details on the required
  changes.

## Usage

```
$ canary.sh -h

canary.sh usage example:

NAMESPACE=books \
  VERSION=v1.0.1 \
  INTERVAL=30 \
  TRAFFIC_INCREMENT=20 \
  DEPLOYMENT=book-ratings \
  SERVICE=book-ratings-loadbalancer \
  canary.sh

These options would deploy version `v1.0.1` of `book-ratings` using the
image found in the previous version of the deployment with an updated
tag, in the `books` namespace, with traffic coming from the
`book-ratings-loadbalancer` service, migrating 20% of traffic at a time
to the new version at 30 second intervals.

Optional variables:
  KUBE_CONTEXT: defaults to currently selected context.
  HEALTHCHECK: path to executable to run instead of basing health
    on pod restarts. The command or script should return 0
    if healthy and anything else otherwise. If nothing is specified,
    pod restarts are used (see Kubernetes docs for probes).
  HPA: name of Horizontal Pod Autoscaler if there's one targeting
    this deployment.
  ON_FAILURE: path to executable to run if the canary healthcheck
    fails and rolls back.
  WORKING_DIR: defaults to $(mktemp -d).

See https://github.com/zacanger/canary.sh for details.
```

## Example

```bash
NAMESPACE=canary-test \
  VERSION=v0.0.2 \
  INTERVAL=60 \
  TRAFFIC_INCEMENT=20 \
  DEPLOYMENT=awesome-app \
  SERVICE=awesome-app \
  HPA=awesome-app \                # optional var
  HEALTHCHECK=/path/to/my/script \ # optional var
  ON_FAILURE=/path/to/script \     # optional var
  KUBE_CONTEXT=context \           # optional var
  WORKING_DIR=$(pwd) \             # optional var
  canary.sh
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    somecompany.app: awesome-app
  # Append version to name
  name: awesome-app-0.0.1
spec:
  selector:
    matchLabels:
      somecompany.app: awesome-app
  replicas: 7
  template:
    metadata:
      labels:
        # Add app label with name and version
        app: awesome-app-0.0.1
        somecompany.app: awesome-app
    spec:
      containers:
      - image: 'some-repo/awesome-app:0.0.1'
        name: awesome-app
        resources:
          limits:
            cpu: 2000m
            memory: 2Gi
          requests:
            cpu: 1000m
            memory: 1Gi
        livenessProbe:
          httpGet:
            path: /.health/alive
            port: 80
        readinessProbe:
          httpGet:
            path: /.health/ready
            port: 80
        ports:
        - containerPort: 80
        env:
          - name: FOO
            value: 'bar'
      restartPolicy: Always
---
apiVersion: v1
kind: Service
metadata:
  labels:
    somecompany.app: awesome-app
    # Add version label
    version: 0.0.1
  name: awesome-app
spec:
  type: LoadBalancer
  sessionAffinity: ClientIP
  sessionAffinityConfig:
    clientIP:
      timeoutSeconds: 21600
  ports:
  - name: https
    port: 443
    targetPort: 80
  selector:
    somecompany.app: awesome-app
---
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  name: awesome-app
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    # Append version to match deployment name if using an HPA
    name: awesome-app-0.0.1
  minReplicas: 7
  maxReplicas: 20
  targetCPUUtilizationPercentage: 70
```

## Contributing

Pull requests and issues are welcome. See
[CONTRIBUTING](./.github/CONTRIBUTING.md) for details.

## Credits and License

Originally forked from <https://github.com/codefresh-io/k8s-canary-deployment>
(MIT licensed). Heavily modified to work without Codefresh, allow more options,
include better logs and usage messages, treat strings safely, allow custom
healthchecks, allow running scripts on failure, work with Horizontal Pod
Autoscalers, and other changes.

[LICENSE (MIT)](./LICENSE.md)


CHANGELOG.md
# Changes


## v0.4.1

* fix: look at what they did to my boy
* docs: changelog

## v0.4.0

* chore: taking this back i guess
* Added Deployment logs upon failure
* remove hpa during canary rollout (#6)
* wait for k8s deploy to be available (#5)
* formatting
* Grokable (#4)
* Revert "canary rewrite (#2)"
* Revert "Grokable (#3)"
* Grokable (#3)
* canary rewrite (#2)
* fixed edge case when current version is a substr of version (#1)
* fix: add a shellcheck disable
* docs: changelog

## v0.3.1

* chore: cleanup
* docs: changelog

## v0.3.0

* chore: cleanup
* docs: clarify where default healthcheck comes from
* docs(readme): fix typo
* docs: changelog

## v0.2.0

* docs(install): example on installing from a tag
* chore: remove some todos
* docs: changelog

## v0.1.2

* fix(release): uncomment
* docs: releasing
* docs: changelog

## v0.1.1

* docs: changelog

## v0.1.0

* docs: update example
* docs: upgrade example for k8s 1.16+
* chore: log cleanup
* docs: update credits section
* chore: revert to hacker theme
* feat: add link
* chore: try tactile
* chore: try a different theme
* chore: add newline to generated file
* Set theme jekyll-theme-minimal
* feat: add working_dir option
* chore: merge in upstream so those commits aren't lost
* chore(makefile): add uninstall, remove phony from install
* docs: note about verifying contents
* docs: cleanup, reorganize, inline examples
* feat: hpas work!
* feat: hpa
* chore: more logging
* feat: add ON_FAILURE hook
* chore: change some variable names
* fix: 2>&1, also docs
* feat: more validation
* docs: note about hpa status
* docs: example and readme cleanup
* fix: custom healthcheck works now
* docs(example): cleanup
* docs: note about gsed
* fix(makefile): add PHONY
* feat: add tests
* chore: remove test file
* chore: clean up misc project files
* chore: remove working_dir option, always use mktemp -d
* chore: remove input_deployment option
* fix: only gnu sed
* fix: disable deleting hpas until backing up works
* docs: add example
* fix: more validation
* fix: remove broken release script

## v0.0.1

* docs: links, better help, etc
* chore: local variables should be lowercase, and better names
* chore: remove gross let style math expressions
* chore: todos
* chore: store basename 0 in a var for usage
* docs: specify _which_ shell
* chore: move to jane org
* chore: progress
* chore: add todos
* feat: short url
* docs: move some comments to the readme
* chore: comment out unused code
* chore: shellcheck disables (revisit later)
* chore: more validation
* chore: add all the stuff
* Update README.md
* Merge pull request #8 from outime/master
* Fixed bug with spaces in cluster names
* Added MIT license
* Merge pull request #7 from rms1000watt/VTN-13084
* Merge pull request #5 from lanmalkieri/master
* Added link to blog post
* Fixed health check pod selection
* Updated Readme
* Changed time step for example
* Simplified example
* Added Dockerfile for GO application
* Fixed yaml indentation
* Dockerization of canary
* Health check for restarts
* Use labels of service to locate deployment
* Added example
* Package canary script into Dockerfile
* package plugin on its own
* Use kubecontext
* Merge pull request #1 from codefresh-io/fix_scale
* Always scale down production if canary is complete
* Revert "Scale old deployment to zero if canaries are complete"
* Scale old deployment to zero if canaries are complete
* Created readme
* Added sample Codefresh file
* Added dockerfile based on kube-help image
* Added wait period, various cleanups
* Fixed tabs
* Scaling up and down


canary.sh
#!/usr/bin/env bash
set -eo pipefail

canarysh_repo='https://github.com/zacanger/canary.sh'
canarysh=$(basename "$0")
canary_deployment=$DEPLOYMENT-$VERSION
working_dir=
# shellcheck disable=SC2153
if [ -n "$WORKING_DIR" ]; then
  working_dir=${WORKING_DIR%/}
else
  working_dir=$(mktemp -d)
fi

# GNU sed only. This is specified in the readme.
if hash gsed 2>/dev/null; then
  _sed=$(which gsed)
else
  _sed=$(which sed)
fi

usage() {
  cat <<EOF
$canarysh usage example:

NAMESPACE=books \\
  VERSION=v1.0.1 \\
  INTERVAL=30 \\
  TRAFFIC_INCREMENT=20 \\
  DEPLOYMENT=book-ratings \\
  SERVICE=book-ratings-loadbalancer \\
  $canarysh

These options would deploy version \`v1.0.1\` of \`book-ratings\` using the
image found in the previous version of the deployment with an updated
tag, in the \`books\` namespace, with traffic coming from the
\`book-ratings-loadbalancer\` service, migrating 20% of traffic at a time
to the new version at 30 second intervals.

Optional variables:
  KUBE_CONTEXT: defaults to currently selected context.
  HEALTHCHECK: path to executable to run instead of basing health
    on pod restarts. The command or script should return 0
    if healthy and anything else otherwise. If nothing is specified,
    pod restarts are used (see Kubernetes docs for probes).
  HPA: name of Horizontal Pod Autoscaler if there's one targeting
    this deployment.
  ON_FAILURE: path to executable to run if the canary healthcheck
    fails and rolls back.
  WORKING_DIR: defaults to \$(mktemp -d).

See $canarysh_repo for details.
EOF

  # Passed 0 or 1 from validate function
  exit "$1"
}

validate() {
  if ! hash kubectl 2>/dev/null; then
    echo "$canarysh: kubectl is required to use this program"
    exit 1
  fi

  # BSD sed doesn't have a --help flag
  $_sed --help >/dev/null 2>&1 || {
    echo "$canarysh: GNU sed is required"
    exit 1
  }

  if [ "${BASH_VERSINFO[0]}" -lt 4 ]; then
    echo "$canarysh: Minimum supported Bash version is 4"
    exit 1
  fi

  # Match --help, -help, -h, help
  if [[ "$1" =~ "help" ]] || [[ "$1" =~ "-h" ]]; then
    usage 0
  fi

  if [ -z "$VERSION" ] || \
    [ -z "$SERVICE" ] || \
    [ -z "$DEPLOYMENT" ] || \
    [ -z "$TRAFFIC_INCREMENT" ] || \
    [ -z "$NAMESPACE" ] || \
    [ -z "$INTERVAL" ]; then
    usage 1
  fi
}

healthcheck() {
  log="[$canarysh ${FUNCNAME[0]}]"
  echo "$log Starting healthcheck"
  h=true

  if [ -n "$HEALTHCHECK" ]; then
    # Run whatever the user provided, check its exit code
    # Set +e so the parent (this script) doesn't exit
    set +e
    "$HEALTHCHECK"
    # shellcheck disable=SC2181
    if [ $? -ne 0 ]; then
      h=false
    fi
    set -e
  else
    # K8s healthcheck
    output=$(kubectl get pods \
      -l app="$canary_deployment" \
      --no-headers)

    echo "$log $output"
    # shellcheck disable=SC2207
    s=($(echo "$output" | awk '{s+=$4}END{print s}'))
    # NAME READY STATUS RESTARTS AGE

    # Restart count. Restarts mean the probes are failing (or other issues,
    # unlhealthy either way).
    # shellcheck disable=SC2128
    if [ "$s" -gt "4" ]; then
      h=false
    fi
  fi

  if [ ! $h == true ]; then
    echo "$log Canary is unhealthy"
    cancel "Healthcheck failed; canceling rollout"
  else
    echo "$log Service healthy"
  fi
}

cancel() {
  log="[$canarysh ${FUNCNAME[0]}]"
  message=$1
  echo "$log $message"

  if [ -n "$ON_FAILURE" ]; then
    # We don't want to break our script if their thing fails
    set +e
    echo "$log Running \$ON_FAILURE"
    "$ON_FAILURE"
    set -e
  fi

  echo "$log $message"
  #TODO this is not the correct way to roll back a k8s deploy,
  #restore previous rs instead and stop changing the deploy name
  echo "$log Restoring original deployment to $prod_deployment"
  kubectl apply \
    --force \
    -f "$working_dir/original_deployment.yml"
  kubectl rollout status "deployment/$prod_deployment"

  echo "$log Removing canary deployment completely"
  kubectl delete deployment "$canary_deployment"

  echo "$log Restoring previous HPA"
  kubectl apply -f "$working_dir/original_hpa.yml"

  exit 1
}

cleanup() {
  log="[$canarysh ${FUNCNAME[0]}]"

  echo "$log Applying new HPA"
  kubectl apply -f "$working_dir/canary_hpa.yml"

  echo "$log Removing previous deployment $prod_deployment"
  kubectl delete deployment "$prod_deployment"

  echo "$log Marking canary as new production"
  kubectl get service "$SERVICE" -o=yaml | \
    $_sed -e "s/$current_version/$VERSION/g" | \
    kubectl apply -f -
}

increment_traffic() {
  log="[$canarysh ${FUNCNAME[0]}]"
  increment=$1
  max_replicas=$2


  prod_replicas=$(kubectl get deployment \
    "$prod_deployment" \
    -o=jsonpath='{.spec.replicas}')

  canary_replicas=$(kubectl get deployment \
    "$canary_deployment" \
     -o=jsonpath='{.spec.replicas}')

  new_prod_replicas=$((prod_replicas-increment))
  # Sanity check
  if [ "$new_prod_replicas" -lt "0" ]; then
    new_prod_replicas=0
  fi

  new_canary_replicas=$((canary_replicas+increment))
  # Sanity check
  if [ "$new_canary_replicas" -ge "$max_replicas" ]; then
    new_canary_replicas=$max_replicas
    new_prod_replicas=0
  fi

  echo "$log Setting canary replicas to $new_canary_replicas"
  kubectl scale --replicas="$new_canary_replicas" "deploy/$canary_deployment"

  echo "$log Setting production replicas to $new_prod_replicas"
  kubectl scale --replicas=$new_prod_replicas "deploy/$prod_deployment"

  # Verify scaling is where we expect
  time=0
  while [ "$(kubectl get deploy "$canary_deployment" -o=jsonpath='{.status.availableReplicas}')" -ne "$new_canary_replicas" ]; do
    sleep 2
    time=$((time+2))
    if [ "$time" -gt "300" ]; then
      cancel "timeout scaling up"
    fi
    echo -n "."
  done
  echo "$log Success:         canary replicas = $new_canary_replicas"
  time=0
  while [ "$(kubectl get deploy "$prod_deployment" -o=jsonpath='{.status.availableReplicas}')" -ne "$new_prod_replicas" ]; do
    sleep 2
    time=$((time+2))
    if [ "$time" -gt "60" ]; then
      cancel "timeout scaling down"
    fi
    echo -n "."
  done
  echo "$log Success: old production replicas = $new_prod_replicas"
}

copy_resources() {
#  # removed this step as the next one takes care of it, and using both creates
#  # a breaking edge case when current version is substr of version.
  log="[$canarysh ${FUNCNAME[0]}]"
#  # Replace old deployment name with new
#  $_sed -Ei -- "s/name\: $prod_deployment/name: $canary_deployment/g" "$working_dir/canary_deployment.yml"
#  echo "$log Replaced deployment name"

  # Replace docker image
  $_sed -Ei -- "s/$current_version/$VERSION/g" "$working_dir/canary_deployment.yml"
  echo "$log Replaced image in deployment"
  echo "$log Production deployment is $prod_deployment, canary is $canary_deployment"

  # $_sed -Ei -- "s/$current_version/$VERSION/g" "$working_dir/canary_hpa.yml"
  $_sed -Ei -- "s/name\: $prod_deployment/name: $canary_deployment/g" "$working_dir/canary_hpa.yml"
  echo "$log Replaced target in HPA"
}

main() {
  log="[$canarysh ${FUNCNAME[0]}]"
  if [ -n "$KUBE_CONTEXT" ]; then
    echo "$log Setting Kubernetes context"
    kubectl config use-context "${KUBE_CONTEXT}"
  else
    echo "$log ---WARNING: running canaries without passing kube_context is dangerous and support for that pattern will soon end---"
  fi
  echo "$log Setting Kubernetes namespace to ${NAMESPACE}"
  { #try
    kubectl get namespace "${NAMESPACE}"
  } || { #catch
    echo "$log namespace not found.  You may need to set your KUBE_CONTEXT"
    echo "$log KUBE_CONTEXT was set to ${KUBE_CONTEXT} for this build"
    echo "$log Aborting"
    exit 1
  }
  kubectl config set-context --current --namespace="${NAMESPACE}"

  echo "$log Getting current version"
  current_version=$(kubectl get service "$SERVICE" -o=jsonpath='{.metadata.labels.version}')

  if [ -z "$current_version" ]; then
    echo "$log No current version found"
    echo "$log Do you have metadata.labels.version set?"
    echo "$log Aborting"
    exit 1
  fi

  if [ "$current_version" == "$VERSION" ]; then
   echo "$log VERSION matches current_version: $current_version"
   exit 0
  fi

  echo "$log Current version is $current_version"
  prod_deployment=$DEPLOYMENT-$current_version

  echo "$log Getting current deployment"
  kubectl get deployment "$prod_deployment" -o=yaml > "$working_dir/canary_deployment.yml"

  echo "$log Backing up original deployment"
  cp "$working_dir/canary_deployment.yml" "$working_dir/original_deployment.yml"

  if [ -n "$HPA" ]; then
    echo "$log Getting current HPA"
    kubectl get hpa "$HPA" -o=yaml > "$working_dir/canary_hpa.yml"
    cp "$working_dir/canary_hpa.yml" "$working_dir/original_hpa.yml"
  fi

  echo "$log Finding current replicas"

  # Copy existing resources and update
  copy_resources
  log="[$canarysh ${FUNCNAME[0]}]"
  starting_replicas=$(kubectl get deployment "$prod_deployment" -o=jsonpath='{.spec.replicas}')
  echo "$log Found replicas $starting_replicas"

  echo "$log Removing previous HPA"
  kubectl delete -f "$working_dir/canary_hpa.yml"

  echo "$log calculating increment..."

  # find increment float, convert to int(floor), and round up to 1 if needed
  pod_increment=$((TRAFFIC_INCREMENT*starting_replicas/100))
  pod_increment=${pod_increment%.*}
  if [ "$pod_increment" -lt "1" ]; then
    pod_increment=1
  fi
  echo "$log pods will be incremented by $pod_increment"
  # Launch one replica first
  $_sed -Ei -- "s#replicas: $starting_replicas#replicas: 1#g" "$working_dir/canary_deployment.yml"
  echo "$log Launching 1 pod with canary"
  kubectl apply -f "$working_dir/canary_deployment.yml"

  echo -n "$log Waiting for canary pod"
  time=0
  #check to make sure everything is set up correctly for our query to return a number
  isNumber='^[0-9]+$'

  while [ -z "$(kubectl get deploy "$canary_deployment" -o=jsonpath='{.status.availableReplicas}')" ]; do
    sleep 2
    time=$((time+2))
    if [ "$time" -gt "600" ]; then
      deployment_log=$(kubectl logs deploy/"$canary_deployment")
      echo "$log deployment log -- $deployment_log"
      cancel "timeout waiting for k8s deploy to build"
    fi
    echo -n "."
  done
  kubectlResult="$(kubectl get deploy "$canary_deployment" -o=jsonpath='{.status.availableReplicas}')"
  if ! [[ $kubectlResult =~ $isNumber ]]; then
    echo "$log kubectl expected a numbered response, instead got:"
    echo "$log $kubectlResult"
    echo "$log Aborting"
    cancel "non-numbered kubectl response"
  fi
  time=0
  while [ "$(kubectl get deploy "$canary_deployment" -o=jsonpath='{.status.availableReplicas}')" -eq 0 ]; do
    sleep 2
    time=$((time+2))
    if [ "$time" -gt "600" ]; then
      cancel "timeout deploying first replica"
    fi
    echo -n "."
  done
  echo ''
  echo "$log Canary target replicas: $starting_replicas"

  healthcheck
  log="[$canarysh ${FUNCNAME[0]}]"

  echo "$log scaling canaries to $starting_replicas by increments of $pod_increment"
  while true; do
    if [ "$(kubectl get deploy "$canary_deployment" -o=jsonpath='{.status.availableReplicas}')" -ge "$starting_replicas" ]; then
      cleanup
      echo "$log Done"
      exit 0
    fi
    increment_traffic "$pod_increment" "$starting_replicas"
    log="[$canarysh ${FUNCNAME[0]}]"
    echo "$log Sleeping for $INTERVAL seconds"
    sleep "$INTERVAL"
    healthcheck
    log="[$canarysh ${FUNCNAME[0]}]"
  done

}

validate "$@"
main
